pkg/whiteflag/white_flag.go

```diff
@@ -0,0 +1,265 @@
+	wfConfirmation := &Confirmation{
+		TailsIncluded:   make([]hornet.Hash, 0),
+		NewAddressState: make(map[string]int64),
+	}
+	for stack.Len() > 0 {
+		if err := ProcessStack(stack, wfConfirmation, visited, newMilestoneIndex); err != nil {
+			return nil, err
+		}
+		// since we first feed the stack the trunk,
+		// we need to make sure that we also examine the branch path.
+		// however, we only need to do it if the branch wasn't visited yet.
+		// the referenced branch transaction could for example already be visited
+		// if it is directly/indirectly approved by the trunk.
+		_, branchVisited := visited[branchHash]
+		if stack.Len() == 0 && !branchVisited {
+			stack.PushFront(branchHash)
+		}
+	}
+
+	return DefaultHasher.TreeHash(wfConfirmation.TailsIncluded), nil
+}
+
+// ProcessStack retrieves the first element from the given stack, loads its bundle and then the trunk and
+// branch transaction of the bundle head. If trunk and branch are both SEPs, already visited or already confirmed,
+// then the mutations from the transaction retrieved from the stack are accumulated to the given Confirmation struct's mutations.
+// This function must be called repeatedly to compute the mutations a white-flag confirmation would create.
+// If the popped transaction was used to mutate the Confirmation struct, it will also be appended to Confirmation.TailsIncluded
+// and it will be removed from the stack. If the head trunk doesn't meet any of the mentioned criteria, it is pushed onto the
+// stack to be the next transaction to be examined on the subsequent ProcessStack() call (same with the branch
+// but only if the trunk wasn't pushed onto the stack). The ledger state must be write locked while this function
+// is getting called in order to ensure consistency.
+func ProcessStack(stack *list.List, wfConf *Confirmation, visited map[string]struct{}, milestoneIndex milestone.Index) error {
+	// load candidate tail tx
+	ele := stack.Front()
+	currentTxHash := ele.Value.(hornet.Hash)
+	cachedTx := tangle.GetCachedTransactionOrNil(currentTxHash)
+	if cachedTx == nil {
+		return fmt.Errorf("%w: candidate tx %s doesn't exist", ErrMissingTransaction, currentTxHash)
+	}
+	defer cachedTx.Release()
+	currentTx := cachedTx.GetTransaction()
+
+	if !currentTx.IsTail() {
+		return fmt.Errorf("%w: candidate tx %s is not a tail of a bundle", ErrMilestoneApprovedInvalidBundle, currentTx.GetTxHash())
+	}
+
+	// load up bundle to retrieve trunk and branch of the head tx
+	cachedBundle := tangle.GetCachedBundleOrNil(currentTx.GetTxHash())
+	if cachedBundle == nil {
+		return fmt.Errorf("%w: bundle %s of candidate tx %s doesn't exist", ErrMissingBundle, currentTx.Tx.Bundle, currentTx.GetTxHash().Trytes())
+	}
+	defer cachedBundle.Release()
+
+	if !cachedBundle.GetBundle().IsValid() || !cachedBundle.GetBundle().ValidStrictSemantics() {
+		return fmt.Errorf("%w: bundle %s is invalid", ErrMilestoneApprovedInvalidBundle, currentTx.Tx.Bundle)
```

Comment:
@muXxer should we panic here? The RFC explicitely states

_In the event that an invalid bundle was encountered in the confirmation traversal, a node's expected behaviour would be to completely stop operations - and log the error - as it would mean the coordinator confirmed something it should not have._

Reply:
Or should we handle that in the solidifier and check for the ErrMilestoneApprovedInvalidBundle error?